pkg/controller/constraint/constraint_controller.go

```diff
@@ -302,104 +302,14 @@ func (r *ReconcileConstraint) Reconcile(ctx context.Context, request reconcile.R
-					if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
-						return reconcile.Result{}, err
-					}
-					currentVapBinding = nil
-				}
-				transformedVapBinding, err := transform.ConstraintToBinding(instance, VAPEnforcementActions)
-				if err != nil {
-					return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, "could not transform constraint to ValidatingAdmissionPolicyBinding")
-				}
 
-				newVapBinding, err := getRunTimeVAPBinding(groupVersion, transformedVapBinding, currentVapBinding)
-				if err != nil {
-					return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, "could not get ValidatingAdmissionPolicyBinding object with runtime group version")
-				}
-
-				if err := controllerutil.SetControllerReference(instance, newVapBinding, r.scheme); err != nil {
-					return reconcile.Result{}, err
-				}
-
-				if currentVapBinding == nil {
-					log.Info("creating vapbinding")
-					if err := r.writer.Create(ctx, newVapBinding); err != nil {
-						return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, fmt.Sprintf("could not create ValidatingAdmissionPolicyBinding: %s", vapBindingName))
-					}
-				} else if !reflect.DeepEqual(currentVapBinding, newVapBinding) {
-					log.Info("updating vapbinding")
-					if err := r.writer.Update(ctx, newVapBinding); err != nil {
-						return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, fmt.Sprintf("could not update ValidatingAdmissionPolicyBinding: %s", vapBindingName))
-					}
-				}
+			requeueAfter, err := r.generateVAPB(ctx, enforcementAction, instance, status)
+			if err != nil {
+				return reconcile.Result{RequeueAfter: requeueAfter}, err
 			}
-			// do not generate vapbinding resources
-			// remove if exists
-			if !generateVAPB && groupVersion != nil {
-				currentVapBinding, err := vapBindingForVersion(*groupVersion)
-				if err != nil {
-					return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, "could not get ValidatingAdmissionPolicyBinding API version")
-				}
-				vapBindingName := fmt.Sprintf("gatekeeper-%s", instance.GetName())
-				log.Info("check if vapbinding exists", "vapBindingName", vapBindingName)
-				if err := r.reader.Get(ctx, types.NamespacedName{Name: vapBindingName}, currentVapBinding); err != nil {
-					if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
-						return reconcile.Result{}, err
-					}
-					currentVapBinding = nil
-				}
-				if currentVapBinding != nil {
-					log.Info("deleting vapbinding")
-					if err := r.writer.Delete(ctx, currentVapBinding); err != nil {
-						return reconcile.Result{}, r.reportErrorOnConstraintStatus(ctx, status, err, fmt.Sprintf("could not delete ValidatingAdmissionPolicyBinding: %s", vapBindingName))
-					}
-				}
+			if requeueAfter != time.Duration(0) {
```

Comment:
because we are requeueing before `cacheConstraint` is called (which adds the constraint to the constraint framework), the constraint will not be enforced at all until the VAP objects are created. These seems unnecessary and will hurt the performance of existing G8r uses.